open_optical_gating/cli/optical_gater_server.py

```diff
@@ -382,44 +372,28 @@ def determine_state(self, frame):
             logger.success("Period determined.")
 
             # Note, passing the new period to the adaptive system is left to the user/app
-            self.stop = True
+            if self.userInteractionPermitted:
+                self.stop = True
+            else:
+                # TODO: JT writes: I think this is what is needed to automatically switch to "sync" mode at this point,
+                # but hopefully CJN can confirm
+                self.frame_num = 0
+                self.state = "sync"
 
-    def adapt_state(self, frame):
+    def adapt_state(self, pixelArray):
         """ Code to run when in "adapt" state.
             Adaptive prospective optical gating mode
             i.e. update reference sequence, while maintaining the same phase-lock.
             In this mode we determine a new period and then align with
             previous periods using an adaptive algorithm.
         """
-        logger.debug("Processing frame in adaptive optical gating mode.")
-
-        # Obtains a minimum amount of buffer frames
-        if self.frame_num < self.settings["frame_buffer_length"]:
-            logger.debug("Not yet enough frames to determine a new period.")
-
-            # Inserts current frame into buffer
-            self.ref_buffer[self.frame_num, :, :] = frame
-
-            # Increases frame number counter
-            self.frame_num += 1
-
-        # Once a suitable number of frames has been buffered,
-        # gets a new period and aligns to the history
-        else:
-            # Obtains a reference period
-            logger.debug("Determining new reference period")
-            # Calculate period from determine_reference_period.py
-            self.ref_frames, self.pog_settings = ref.establish(
-                self.ref_buffer, self.pog_settings
-            )
-
-            # Determine barrier frames
-            self.pog_settings = pog.determine_barrier_frames(self.pog_settings)
-
-            # Save the period
-            ref.save_period(self.ref_frames, self.settings["period_dir"])
-            logger.success("Period determined.")
+        
+        # Start by calling through to determine_state() to establish a new reference sequence
+        # TODO: JT writes: CJN can you confirm whether we should be setting self.stop (which determine_state will do)
+        # when we are in "adapt" mode like this?
+        self.determine_state(pixelArray, modeString="adaptive optical gating")
 
+        if self.ref_frames is not None:
             self.state = "sync"
 
             self.frame_num = 0
```

Comment:
Important: as far as I can see, self.global_solution is never initialised anywhere. Not sure what has gone wrong here, but this means that LTU is broken.

Reply:
Replying to your top-level comment which is responding to this: I hadn't been sure how much you had tested these changes (since I encountered a number of bugs along the way). If you've tested these changes, it works for your FileOpticalGater, and that is definitely doing a LTU (when I ran it, it looked as if it might not have done one...?), then this must be something to do with the fact that I am testing it "truly live" rather than with manual state changes driven by FileOpticalGater.

If that's the case, I'll dig deeper and try and work out what the root problem is. You could be right about the None, I'm not sure right now.

Reply:
... although when I am running it with an unbroken stream of input images (what I rather misleadingly referred to as "truly live") then the issue is that the variable simply does not exist when making the function call, and so the code fails. That does not surprise me, because I can't see anywhere whatsoever that it is initialized, in anybody's code. But feel free to tell me I have missed something!

Reply:
OK right, self.global_solution is set up in user_select_period(). So I guess I just need to provide alternative functionality to that, for when we are running unattended. But we can discuss tomorrow.